### PR TITLE
Refactor parameter mechanism

### PR DESCRIPTION
**Pull request by xiaoling-yi**
In this PR, we
* optimize the parameter passing mechanism. Specifically, we use **case classes** to organize the parameters for each sub-module and streamer. We use a **trait** to organize the inferred parameters for Streamer to remove some duplicated code for inferred parameter generation. Address suggestion (1) in https://github.com/KULeuven-MICAS/snax-streamer/pull/2
* a common case class for both Data Reader and Writer
* add comments for the new parameter mechanism
* related parameter changes to tests
* rename dataReader/WriterTcdmPorts -> tcdmPortsNum, genSpatialLoopIndeces-> genSpatialLoopIndices.
Sorry for the large number of line changes!!

### CODE REVIEW

**Review by jorendumoulin — APPROVED** [review]
These are some beautiful changes 😍 

My main comment is that I would move all parameter classes to the `Params.scala` file. This way, they are all together and it will be very clear what all the different parameters are!

File: src/main/scala/streamer/DataReader.scala [hunk 1]
```diff
@@ -3,37 +3,57 @@ package streamer
...
+  * @param elementWidth
+  *   single data element width for each data mover, useful for generating
+  *   unrolling addresses
+  * @param fifoWidth
+  *   FIFO width
+  */
+case class DataMoverParams(
+    tcdmPortsNum: Int = DataMoverTestParameters.tcdmPortsNum,
+    unrollingFactor: Seq[Int] = DataMoverTestParameters.unrollingFactor,
+    unrollingDim: Int = DataMoverTestParameters.unrollingDim,
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply src/main/scala/streamer/DataReader.scala:25]: Good! I will change that.

**Review by JosseVanDelm — APPROVED** [review]
Overall this looks quite nice!
```scala
** This class represents all the parameters for the Data Mover (including Data
  * Reader and Data Writer).
  *
  * @param tcdmPortsNum
  *   the number of TCDM ports connected to each data mover
  * @param spatialBounds
  *   spatial unrolling factors (your parfor) for each data mover
  * @param spatialDim
  *   the dimension of spatial unrolling factors (your parfor) for each data
  *   mover
  * @param elementWidth
  *   single data element width for each data mover, useful for generating
  *   unrolling addresses
  * @param fifoWidth
  *   FIFO width
  */
```
Can we get some more of these beautiful ^^ comments in future PRs please :pray:

**Comment by xiaoling-yi** [comment]
@jorendumoulin , thanks for your advice! 

> Overall this looks quite nice!
> 
> ```scala
> ** This class represents all the parameters for the Data Mover (including Data
>   * Reader and Data Writer).
>   *
>   * @param tcdmPortsNum
>   *   the number of TCDM ports connected to each data mover
>   * @param spatialBounds
>   *   spatial unrolling factors (your parfor) for each data mover
>   * @param spatialDim
>   *   the dimension of spatial unrolling factors (your parfor) for each data
>   *   mover
>   * @param elementWidth
>   *   single data element width for each data mover, useful for generating
>   *   unrolling addresses
>   * @param fifoWidth
>   *   FIFO width
>   */
> ```
> 
> Can we get some more of these beautiful ^^ comments in future PRs please 🙏

ok!

**Pull request merged**